clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java

```diff
@@ -972,16 +975,28 @@ private Record nextFetchedRecord() {
         private List<ConsumerRecord<K, V>> fetchRecords(int maxRecords) {
             if (isFetched)
                 return Collections.emptyList();
+            if (nextInlineException != null) {
```

Comment:
@hachikuji OK misunderstood your previous example. Yes, since record 1 is invalid, the offset will remain at 1. However, there is same problem as of current Kafka implementation without this patch. I think fixing that behavior is out of the scope of this patch.